flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java

```diff
@@ -1090,6 +1104,26 @@ else if (current == CANCELED || current == FAILED) {
 		}
 	}
 
+	private void finishPartitionsAndScheduleOrUpdateConsumers() {
+		final List<IntermediateResultPartition> newlyFinishedResults = getVertex().finishAllBlockingPartitions();
+		if (newlyFinishedResults.isEmpty()) {
+			return;
+		}
+
+		final HashSet<ExecutionVertex> consumersToSchedule = new HashSet<>();
```

Comment:
One side effect of using a Set is that the vertices are disordered.
It's not problematic but may make the scheduling logs a bit mess.

e.g. A has 2 downstream JobVertices B, C, when all instances of A finishes
previously it is:

> ... B1 ... transitioned from CREATED to SCHEDULED
> ... B2 ... 
> ...
> ... Bn ... 
> ... C1 ... 
> ... C2 ... 
> ...
> ... Cn ... transitioned from CREATED to SCHEDULED

With this change, it might be:

> ... B2 ... transitioned from CREATED to SCHEDULED
> ... Cn ... 
> ...
> ... B3 ... 
> ... C5 ... 
> ... C2 ... 
> ...
> ... B1 ... transitioned from CREATED to SCHEDULED

We can avoid the disorders by adding the vertices to schedule in a list and using the set to track vertices added. But I'm not sure whether it's needed.
WDYT?

Reply:
Okay, sure, we can a set to just detect duplicates, and still schedule them in order.